enhancements/multi-arch/dynamic-imagestream-importmode-setting.md

```diff
@@ -0,0 +1,469 @@
+value.
+4. openshift-apiserver-operator updates the
+`OpenshiftAPIServer`'s observed config ImagePolicyConfig
+field with the import mode from the image config status.
+5. openshift-apiserver uses the value of the
+import mode in the observed config's
+ImagePolicyConfig field to set the default import
+mode for any newly created imagestream.
+
+Note that a change to the import mode in the image
+config CR would trigger a redeployment of the
+openshift-apiserver
+
+### API Extensions
+
+- image.config.openshift.io: Add
+  `ImagestreamImportMode` config flag to CRD spec
+  and status.
+
+``` 
+type ImageSpec struct {
+...
+...
+// imagestreamImportMode controls the import
+// mode behaviour of imagestreams. It can be set
+// to `Legacy` or `PreserveOriginal`. The
+// default behaviour is to default to Legacy.
+// If this value is specified, this setting is
+// applied to all imagestreams which do not
+// have the value set.
+// +optional
+ImageStreamImportMode imagev1.ImportModeType
+`json:"imagestreamImportMode,omitempty" }`
+```
+```
+type ImageStatus struct {
+...
+...
+// imagestreamImportMode controls the import
+// mode behaviour of imagestreams. It can be set
+// to `Legacy` or `PreserveOriginal`. The
+// default behaviour is to default to Legacy.
+// If this value is specified, this setting is
+// applied to all new imagestreams which do not
+// have the value set.
+// +optional
+ImagestreamImportMode imagev1.ImportModeType
+`json:"imagestreamImportMode,omitempty"
+}
+```
+- openshiftcontrolplane/v1/: Add
+  `ImagestreamImportMode` config flag to
+  ImagePolicyConfig struct.
+
+```
+type ImagePolicyConfig {
+...
+...
+// imagestreamImportMode controls the import
+mode behaviour of imagestreams. It can be set
+to `Legacy` or `PreserveOriginal`. The
+// default behaviour is to default to Legacy.
+// If this value is specified, this setting is
+// applied to all imagestreams which do not
+// have the value set.
+// +optional
+ImagestreamImportMode imagev1.ImportModeType
+`json:"imagestreamImportMode,omitempty"`
+
+}
+```
+- An example CR for `image.config.openshift.io`
+  would look like below if the import mode is set:
+```
+apiVersion: config.openshift.io/v1
+kind: Image
+metadata:
+  name: cluster
+  ...
+  ...
+spec:
+  imagestreamImportMode: PreserveOriginal
+status:
+   internalRegistryHostname: image-registry.openshift-image-registry.svc:5000
+   imagestreamImportMode: PreserveOriginal
+  ...
+  ...
+
+```
+- The ClusterVersion status field CR would look
+  like below after CVO has inferred the payload
+  type:
+```
+apiVersion: v1
+items:
+- apiVersion: config.openshift.io/v1
+  kind: ClusterVersion
+  spec:
+    channel: candidate-4.15
+    desiredUpdate:
+      force: false
+      version: 4.15.0
+  status:
+    availableUpdates:
+    ...
+    capabilities:
+    ...
+    conditions:
+    ...
+    desired:
+    ...
+    history:
+    ...
+    payloadArchitecture: amd64  # could be arm64/ppc64le/s390x/multi
```

Comment:
The current `spec.desiredUpdate` property is [`architecture`][1], and the valid values are [`Multi` or empty][2].  We'd initially floated `multi` to match the Cincinnati query parameter, but [were told to use CamelCase for consistency with other Kube enums][3].

It would be theoretically possible to also pass along the single-arch payload architecture, but I don't see why that would be necessary (components in a single-arch cluster that need to know can use [`GOARCH`][5] or similar), and the [`release.openshift.io/architecture` metadata is `multi` or unset][4].  So I expect we want:

```
desired:
  architecture: Multi  # optional, unset/empty-string if it's single arch
```

here just for consistency.  And that we want it under `desired` to mirror the `spec.desiredUpdate` nesting.

[1]: https://github.com/openshift/api/blob/ac9356ba178405e39af0a76adc7d20a529bd93af/config/v1/types_cluster_version.go#L654-L664
[2]: https://github.com/openshift/api/blob/ac9356ba178405e39af0a76adc7d20a529bd93af/config/v1/types_cluster_version.go#L281-L289
[3]: https://github.com/openshift/api/pull/1339#discussion_r1019591915
[4]: https://github.com/openshift/enhancements/blob/1418157429ac0e23855f1bb5bfeda3ba756281bd/enhancements/multi-arch/heterogeneous-architecture-clusters.md?plain=1#L173-L175
[5]: https://pkg.go.dev/runtime#pkg-constants

Reply:
makes sense. I will the doc accordingly

Reply:
updated.